video-encoder-stats: add video encoder stats elements

diegonieto opened the pull request:
- `video-encoder-stats`:
  The element that collects statistics from a video encoder, and attach them onto the `GstBuffers` as metadata. It helps analyze encoding performance and quality metrics.

- `video-compare-mixer`:
  The element in charge of comparing and mixing multiple video streams. Useful for side-by-side quality comparisons or blending outputs from different encoders.

- `videoencoderstatsmeta`:
  Defines metadata structures and logic for handling video encoder statistics along the pipeline. It has been defined as `GstVideoEncoderStatsMetaAPI`.

Running example: `cargo r --example video-encoder-stats`

<img width="1280" height="572" alt="encoder-stats" src="https://github.com/user-attachments/assets/9af04783-9915-48d0-8eda-7fb9d7e7e461" />

diegonieto committed:
videostats: add video encoder stats elements

diegonieto committed:
video-encoder-stats: add vmaf metrics

diegonieto committed:
video-compare-stats: show half decoded outputs for mixing

diegonieto committed:
video-encoder-stats: fix issue for different encoders

Different encoders introduce different latencies and the pipelines where
not starting properly.
Introduce multiples queues in bifurcations and convergences.
Introduce also videoconvert and fix the video caps format to I420

diegonieto committed:
video-compare-mixer: set CPU as default mixer

diegonieto committed:
video-compare-stats: init element without the need of specify a backend

Change start-up sequence

diegonieto committed:
video-compare-mixer: add property to allow split the screen or duplicate

diegonieto committed:
video-compare-mixer: handle crop for glvideomixer

diegonieto committed:
video-encoder-stats: allow to specify the internal decoder

diegonieto committed:
video-encoder-stats: implement request pad for decoded data

This adds a new pad allowing to directly connect to the decoded
data.
It also introduce both videoencoderstats and videocomparemixer unit
tests.

diegonieto committed:
video-encoder-stats: calculate score every framerate frames

This reduces drastically the processing time taken by the vmaf
element, avoiding to post scores for each frame.

diegonieto committed:
videostats: add originalbuffer internal dependency for building

diegonieto committed:
encoderstats: add max. internal buffers metric

diegonieto committed:
videostats: update data layout

diegonieto committed:
encoderstats: fix name causing problems CPU stats

diegonieto committed:
videostats: fix originalbuffer version

diegonieto committed:
video-compare-mixer: add navigation events

diegonieto committed:
video-encoder-stats: create vmaf property and handle optional behaviour

* Detect in runtime whether vmaf element is available
* Check whether property is enabled. Default: enabled
* Expose decode_src pad only if vmaf pipeline branch is enabled
* Add a new example not using vmaf

diegonieto committed:
video-encoder-stats: handle input queues internally

diegonieto committed:
video-encoder-stats: add latency metric

* get latency from SystemTime
** TODO: implement npt as webrtc-precise-sync-send.rs does
* add timestamp before encoding as part of videoencoderstatsmeta
* compare the input timestamp in video-compare-mixer with the current
timestamp (after decoding the video stream)
* add a new example video-stats-zero-latency to showcase the latency
difference

diegonieto committed:
video-encoder-stats: let vmaf score a N/A by default

If vmaf metric is not available N/A will be shown in the metric instead of 0 value.

diegonieto committed:
video-encoder-stats: add silent property and post stats as messages

This enables posting the stats as messages unless the new property silent is set true, by default is false

diegonieto committed:
videoencoderstats: update metric output as average process time

Remove throughput metric and leave avg. processing time instead

diegonieto committed:
video-encoder-stats: update meta timestamps of the current buffer instead of global values

Relates to RDI-2963

rgonzalezfluendo commented on video/stats/src/videoencoderstats.rs:
```diff
@@ -0,0 +1,189 @@
+        writeln!(
+            f,
+            "Output size: {} KB",
+            self.num_bytes / 1024, // Convert to KB
```
```suggestion
            self.num_bytes / 1000, // Convert to KB
```

note diff between KiB and KB.

```
gst-launch-1.0 --gst-plugin-path=$(realpath gstreamer-1.0) -v videotestsrc num-buffers=100 ! video/x-raw,width=1280,height=720,framerate=30/1 ! video-encoder-stats encoder="flulcevch264enc bitrate=5000 keyframe-period=30 max-async-frames=9" ! filesink location=/tmp/caca

/GstPipeline:pipeline0/GstEncoderStats:encoderstats0: last-message = Encoder: flulcevch264enc
Output size: 2495 KB
Num buffers: 100
Bitrate: 5988.720 kbps
Processing time: 66 ms
CPU: 0.42 s
VMAF: N/A
Encode latency: 68.107 ms
End2End latency: 68.996 ms

Got EOS from element "pipeline0".
Execution ended after 0:00:00.813719643
Setting pipeline to NULL ...
Freeing pipeline ...
rgonzalez@rgonzalez-ThinkPad-T14:~/tmp/6g-xr/RDI-2957 $ ls -l /tmp/caca
-rw-r--r-- 1 rgonzalez rgonzalez 2555187 Sep 18 17:34 /tmp/caca
rgonzalez@rgonzalez-ThinkPad-T14:~/tmp/6g-xr/RDI-2957 $ ls -lh /tmp/caca
-rw-r--r-- 1 rgonzalez rgonzalez 2.5M Sep 18 17:34 /tmp/caca
rgonzalez@rgonzalez-ThinkPad-T14:~/tmp/6g-xr/RDI-2957 $ python
Python 3.13.7 (main, Aug 15 2025, 12:34:02) [GCC 15.2.1 20250813] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 2555187/1024
2495.2998046875
```

Linux uses KB and MB (not KiB and MiB)

rgonzalezfluendo commented on video/stats/src/videoencoderstats.rs:
```diff
@@ -0,0 +1,189 @@
+        } else {
+            0.0
+        };
+        let bitrate_str = bitrate/1024.0; // Convert to kbps
```
```suggestion
        let bitrate_str = bitrate/1000.0; // Convert to kbps
```

BIS

rgonzalezfluendo commented on video/stats/src/videoencoderstats.rs:
```diff
@@ -0,0 +1,189 @@
+            encode_latency
+        )?;
+
+        let current_time = gst::SystemClock::obtain();
+        let latency = (current_time.time().unwrap().nseconds() as f64 - self.input_time as f64) / 1_000_000.0;
+
+        writeln!(
+            f,
+            "End2End latency: {:.3} ms",
+            latency
+        )
```
IMHO I would drop this. it is not End2End latency and IMHO it doesn't add any useful information because we can not assume input_queue is after the source.

Also it out of the scope of `video-encoder-stats`.

diegonieto replied:
true, removed here: https://github.com/fluendo/gst-plugins-rs/pull/4/commits/4826fdcdf9adae3733c0939a957935fb9dd80720

rgonzalezfluendo commented on video/stats/src/videoencoderstats.rs:
```diff
@@ -0,0 +1,189 @@
+
+        let pre_encode_time = &self.pre_encode_time;
+        let post_encode_time = &self.post_encode_time;
+        let encode_latency = (*post_encode_time as f64 - *pre_encode_time as f64) / 1_000_000.0;
```
Encode latency is almost the same as processing time. The processing time is the average of all buffers, and encode latency is for each buffer, right?

But I did this change and the values are different. Strange...

<details>
 <summary>patch</summary>

```
diff --git a/video/stats/src/encoderstats/imp.rs b/video/stats/src/encoderstats/imp.rs
index a0c56e39..e834ee29 100644
--- a/video/stats/src/encoderstats/imp.rs
+++ b/video/stats/src/encoderstats/imp.rs
@@ -152,6 +152,7 @@ impl EncoderStats {
             if let Some(mut meta) = buffer.meta_mut::<VideoEncoderStatsMeta>() {
                 let mut new_stats = meta.stats().clone();
                 new_stats.post_encode_time = current_time;
+                stats.lock().unwrap().total_processing_time_2 += (new_stats.post_encode_time - new_stats.pre_encode_time);
 
                 meta.replace(new_stats);
 
diff --git a/video/stats/src/videoencoderstats.rs b/video/stats/src/videoencoderstats.rs
index f15ababe..24a97fd6 100644
--- a/video/stats/src/videoencoderstats.rs
+++ b/video/stats/src/videoencoderstats.rs
@@ -34,6 +34,7 @@ pub struct VideoEncoderStats {
     pub time_last_buffers: VecDeque<Instant>,
     pub max_buffers_inside: usize,
     pub total_processing_time: Duration,
+    pub total_processing_time_2: u64,
     pub threads_utime: u64,
     pub threads_stime: u64,
     pub framerate: Option<gst::Fraction>,
@@ -53,6 +54,7 @@ impl Default for VideoEncoderStats {
             time_last_buffers: VecDeque::<Instant>::new(),
             max_buffers_inside: 0,
             total_processing_time: Duration::ZERO,
+            total_processing_time_2: 0,
             threads_utime: 0,
             threads_stime: 0,
             vmaf_score: None,
@@ -116,8 +118,13 @@ impl fmt::Display for VideoEncoderStats {
         };
         let bitrate_str = bitrate/1024.0; // Convert to kbps
 
+        writeln!(f, "Num buffers: {}", self.num_buffers)?;
         writeln!(f, "Bitrate: {:.3} kbps", bitrate_str)?;
 
+        writeln!(f, "total_processing_time 1: {:?}", self.total_processing_time)?;
+        writeln!(f, "total_processing_time 2: {:?}", self.total_processing_time_2)?;
+
+
         let avg_processing_time = self.avg_processing_time().as_millis();
         writeln!(
             f,
```

</details>


For instance `25.184094281s` vs `25.150373246s`
```
/GstPipeline:pipeline0/GstEncoderStats:encoderstats0: last-message = Encoder: x265enc
Output size: 986 KB
Num buffers: 100
Bitrate: 2367.105 kbps
total_processing_time 1: 25.184094281s
total_processing_time 2: 25150373246
Processing time: 251 ms
CPU: 5.22 s
VMAF: N/A
Encode latency: 1032.306 ms
End2End latency: 1033.626 ms
``

diegonieto replied:
really weird, I've been running this pipeline:
```
gst-launch-1.0 souphttpsrc location=\"https://ftp.nluug.nl/pub/graphics/blender/demo/movies/ToS/tears_of_steel_1080p.mov\" ! qtdemux name=demux demux.video_0 ! queue ! decodebin3 ! videoconvertscale ! capsfilter caps=\"video/x-raw,aspect-ratio=1/1\" ! tee name=tee ! video-encoder-stats encoder=\"x264enc bitrate=1024 tune=zerolatency speed-preset=ultrafast threads=4\" name=vs0 vmaf-stats=false tee. ! video-encoder-stats encoder=\"x264enc bitrate=512 tune=zerolatency speed-preset=ultrafast\" name=vs1 video-compare-mixer split-screen=true backend=CPU name=mixer vs0.src ! h264parse ! avdec_h264 ! mixer.sink_0  vs1.src ! h264parse ! avdec_h264 ! mixer.sink_1  mixer. ! autovideosink -v
```

with your patch and I see almost identical values after a long run

```
/GstPipeline:pipeline0/GstEncoderStats:vs1: last-message = Encoder: x264enc
Output size: 8280 KB
Num buffers: 3570
Bitrate: 445.347 kbps
total_processing_time 1: 25.290121055s
total_processing_time 2: 25292717208
Processing time: 7 ms
CPU: 47.68 s
VMAF: N/A
Encode latency: 6.069 ms
End2End latency: 7.683 ms

/GstPipeline:pipeline0/GstEncoderStats:vs0: last-message = Encoder: x264enc
Output size: 16715 KB
Num buffers: 3570
Bitrate: 898.989 kbps
total_processing_time 1: 22.632750704s
total_processing_time 2: 22638976626
Processing time: 6 ms
CPU: 47.55 s
VMAF: N/A
Encode latency: 7.141 ms
End2End latency: 9.407 ms
```
and both encode latency is in the range of the processing time, and the total processing time match :thinking:

diegonieto replied:
I've been now comparing x265 vs x264:
```
/GstPipeline:pipeline0/GstEncoderStats:vs1: last-message = Encoder: x264enc
Output size: 8165 KB
Num buffers: 3446
Bitrate: 454.979 kbps
total_processing_time 1: 27.792252549s
total_processing_time 2: 27797596912
Processing time: 8 ms
CPU: 44.47 s
VMAF: N/A
Encode latency: 7.226 ms

/GstPipeline:pipeline0/GstEncoderStats:vs0: last-message = Encoder: x265enc
Output size: 14688 KB
Num buffers: 3446
Bitrate: 818.388 kbps
total_processing_time 1: 75.500975765s
total_processing_time 2: 75505265723
Processing time: 21 ms
CPU: 261.13 s
VMAF: N/A
Encode latency: 22.058 ms
```
I got again almost the same total processing time in both cases. Can you share your pipeline, so I can try to replicate? 

For long runs (143s) I can explain the total processing time difference (5ms) can be due to avg. processing time is set before pre and post encoding timestamps, and the sooner requires a lock to sync with global:
```
            stats.lock().unwrap().buffer_out();
            gst::log!(CAT, "Buffer out encoder src pad");

            let current_time = *gst::ClockTime::from_nseconds(
                gst::SystemClock::obtain().upcast::<gst::Clock>().time().unwrap().nseconds()
            );

            let buffer = buffer.make_mut();
            if let Some(mut meta) = buffer.meta_mut::<VideoEncoderStatsMeta>() {
                let mut new_stats = meta.stats().clone();
                new_stats.post_encode_time = current_time;
                stats.lock().unwrap().total_processing_time_2 += (new_stats.post_encode_time - new_stats.pre_encode_time);

                meta.replace(new_stats);

                gst::log!(CAT, "Buffer out encoder src pad, post_encode_time: {}", current_time);
            } else {
                gst::warning!(CAT, "No VideoEncoderStatsMeta found on buffer in encoder src probe");
            }
```
Any thoughts about this?

rgonzalezfluendo commented on video/stats/examples/video-stats-split-screen.rs:
```diff
@@ -0,0 +1,34 @@
+
+fn main() -> Result<(), Error> {
+    gst::init()?;
+
```
Needed to avoid `Error:` no element "video-encoder-stats`
```suggestion

`gstvideostats::plugin_register_static().expect("Failed to register videostats plugin");`
```


Note: https://github.com/fluendo/gst-plugins-rs/blob/main/utils/togglerecord/examples/gtk_recording.rs#L326

diegonieto replied:
true. Added here for all the examples: https://github.com/fluendo/gst-plugins-rs/pull/4/commits/80dea0b6b6337209d8659151494095e904e81be3

rgonzalezfluendo commented on video/stats/src/comparemixer/imp.rs:
```diff
@@ -0,0 +1,779 @@
+        // FIXME remove split_screen logic if not needed. It adds and links crops always
+        self.link_elements(&compositor, true, backend)?;
+
+        self.add_overlay_probe(&self.overlay0);
+        self.add_overlay_probe(&self.overlay1);
```
I don't like to have two different overlays. The UX with only one is much better.

Compare

<img width="1918" height="797" alt="image" src="https://github.com/user-attachments/assets/2b6bf549-6fea-4dce-95f3-8f8a934790c5" />
or 
<img width="1918" height="797" alt="image" src="https://github.com/user-attachments/assets/6d52dfa9-a169-4d93-bd7f-96cfe3842077" />

with

<img width="1268" height="712" alt="image" src="https://github.com/user-attachments/assets/3737b0eb-7216-48fc-aa6a-297d6122eeda" />
or
<img width="1268" height="712" alt="image" src="https://github.com/user-attachments/assets/18010b30-27c4-4d9c-8281-818729ece8b9" />


Also IMHO the overlay must be out of the `video-compare-mixer`. Application must do the overlay.

diegonieto replied:
I understand your point. I wanted to make two of them to control independently, but it's true if we are comparing we will probably show the same parameters and so on. Also you approach will improve the performance. As it implies a refactor I will create a specific task for this: https://fluendo.atlassian.net/browse/RDI-2975

diegonieto replied:
I've unified the overlay here: https://github.com/fluendo/gst-plugins-rs/pull/4/commits/8010eea9a9164d2e4adf1bed8fdeb5112aa0bef1

diegonieto committed:
Update video/stats/src/videoencoderstats.rs

Co-authored-by: Ruben Gonzalez <56379722+rgonzalezfluendo@users.noreply.github.com>

diegonieto committed:
Update video/stats/src/videoencoderstats.rs

Co-authored-by: Ruben Gonzalez <56379722+rgonzalezfluendo@users.noreply.github.com>

diegonieto committed:
videocomparestats: remove end to end latency

It might not represent the actual end2end latency

diegonieto committed:
videostats: add plugin register in all the examples

diegonieto commented on video/stats/src/encoderstats/imp.rs:
```diff
@@ -0,0 +1,829 @@
+            let num_buffers = stats.num_buffers;
+
+            if num_buffers % (fps_n as u64) == 0 {
+                let queue_name = if obj_name.contains("0") { "encq0:src" } else { "encq1:src" };
```
@rgonzalezfluendo I am measuring the CPU time based on the time spent on the thread (queue) that contains the encoder + encoder specific threads. I see x265 CPU is really high compared with x264, but it can be normal. What do you think?
```
/GstPipeline:pipeline0/GstEncoderStats:vs1: last-message = Encoder: x264enc
Output size: 8165 KB
Num buffers: 3446
Bitrate: 454.979 kbps
total_processing_time 1: 27.792252549s
total_processing_time 2: 27797596912
Processing time: 8 ms
CPU: 44.47 s
VMAF: N/A
Encode latency: 7.226 ms

/GstPipeline:pipeline0/GstEncoderStats:vs0: last-message = Encoder: x265enc
Output size: 14688 KB
Num buffers: 3446
Bitrate: 818.388 kbps
total_processing_time 1: 75.500975765s
total_processing_time 2: 75505265723
Processing time: 21 ms
CPU: 261.13 s
VMAF: N/A
Encode latency: 22.058 ms
```

diegonieto committed:
videostats: update to the latest gst-plugins-rs bindings

Fix time object new methods

diegonieto committed:
video-compare-mixer: unify overlay into a single element

videoencoderstats meta is now read from the input queues. The overlay is
not linked after the compositor and the values are updated once per
second according to the given framerate of the pipeline

A new fmt for video-compare-mixer has been created to compose both stats
left and right into a single text for the overlay

diegonieto committed:
video-compare-mixer: add "overlay-stats" property to control overlay

In this way it is possible to disable the overlay of the stats in the
screen. It defaults to enabled.

diegonieto committed:
video-compare-mixer: remove deprected code

diegonieto committed:
video-encoder-stats: remove println!

diegonieto committed:
video-compare-mixer: improve layout alignment and polish code

diegonieto committed:
videostats: update license and description

diegonieto commented on video/stats/src/comparemixer/imp.rs:
```diff
@@ -0,0 +1,905 @@
+}
+
+impl fmt::Display for QueueStats {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```
I would need your help here @rgonzalezfluendo  to properly align the text in the layout

rgonzalezfluendo replied:
What's the problem?

diegonieto committed:
video-encoder-stats: isolate encoder GstObject between two queues

Before the encoder was mixed with another objects that may introduce a
little overhead. This way we ensure only encoder CPU is taken into
account

diegonieto committed:
videostats: merge examples into a single one

diegonieto committed:
video-compare-mixer: fix alignment issue

diegonieto committed:
video-encoder-stats: provide a stats message when EOS

Silent intermediate messages and signals based on property value.
Provide always messages and signal with the final stats on EOS

diegonieto committed:
videoencoderstats: provide Max. buffers inside metric

diegonieto committed:
videoencoderstats: add num buffers needed for the first output metric

diegonieto committed:
videoencoderstats: add current buffers seens metric

msabiniok committed:
video-encoder-stats: handle fractional framerate cases

rgonzalezfluendo committed:
video-encoder-stats: Fixes UX for "Num. Buffers first output"

From
```
Num. Buffers first output: -1
Num. Buffers first output: 1
Num. Buffers first output: 1
```

to

```
Num. Buffers first output: No first output yet
Num. Buffers first output: 1
Num. Buffers first output: 1
```

Issue: RDI-3210

rgonzalezfluendo committed:
video-encoder-stats: Add max_processing_time

This maximum value is only valid if there is no reordering of frames
within the encoder (b-frames). If reordering occurs, it is necessary
to mark the buffers to relate the input buffer to the output buffer.
Similar to MR9622

MR9622: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/9622

Issue: RDI-3209

rgonzalezfluendo committed:
videoencoderstats: fmt
